src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file methods/xgboost/xgboost_tree_regressor.hpp
+ * @author Rishabh Garg
+ *
+ * Definition of the xgboost tree regressor class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_XGBOOST_TREE_REGRESSOR_HPP
+#define MLPACK_METHODS_XGBOOST_XGBOOST_TREE_REGRESSOR_HPP
+
+#include <mlpack/methods/decision_tree/decision_tree_regressor.hpp>
+#include <mlpack/methods/decision_tree/multiple_random_dimension_select.hpp>
+
+namespace mlpack {
+namespace ensemble {
+
+/**
+ * The XGboostTreeRegressor class provides the implementation of Gradient
+ * Boosted Decision Tree Regressor as described in the XGBoost paper:
+ *
+ * @code
+ * @inproceedings{Chen:2016:XST:2939672.2939785,
+ *   author = {Chen, Tianqi and Guestrin, Carlos},
+ *   title = {{XGBoost}: A Scalable Tree Boosting System},
+ *   booktitle = {Proceedings of the 22nd ACM SIGKDD International Conference on
+ *                Knowledge Discovery and Data Mining},
+ *   series = {KDD '16},
+ *   year = {2016},
+ *   isbn = {978-1-4503-4232-2},
+ *   location = {San Francisco, California, USA},
+ *   pages = {785--794},
+ *   numpages = {10},
+ *   url = {http://doi.acm.org/10.1145/2939672.2939785},
+ *   doi = {10.1145/2939672.2939785},
+ *   acmid = {2939785},
+ *   publisher = {ACM},
+ *   address = {New York, NY, USA},
+ *   keywords = {large-scale machine learning},
+ * }
+ * @endcode
+ */
+template<typename LossFunction = SSELoss,
+         typename DimensionSelectionType =
+             mlpack::tree::MultipleRandomDimensionSelect,
+         template <typename> class NumericSplitType = XGBExactNumericSplit>
```

Comment:
Should we allow specifying the categorical split type here?  Below I see you are always using `AllCategoricalSplit`, which I *think* will work, but it could be nice to have it user-specifiable too.

Reply:
Actually, I did think about it. I have made the API considering that we will be having support for categorical features too in the future. When we add it in future, all we will need is to add one constructor, one public Train overload and a template parameter for it.

As of now, I didn't add a template parameter for it because we do not support it yet and it will cause unnecessary confusion.